### PR TITLE
Debug_stacktrace

### PR DESCRIPTION
**Pull request by EmileTrotignon**
This add stacktraces to the html debug output. The stacktrace is only present on hover.

### CODE REVIEW

**Comment by EmileTrotignon** [comment]
image

File: lib/Conf.ml [hunk 1]
```diff
@@ -1401,7 +1401,9 @@ module Operational = struct
   let debug =
     let doc = "Generate debugging output." in
     Decl.flag ~default ~names:["g"; "debug"] ~doc ~kind
-      (fun conf elt -> update conf ~f:(fun f -> {f with debug= elt}))
+      (fun conf elt ->
+        if elt.v then Box_debug.enable_stacktraces := true ;
```
Comment (Julow) [inline lib/Conf.ml:1405]: This means that enabling the debug output implies enabling the stacktraces. Is the purpose to have the debugging output without having the stacktraces ?

File: lib/Fmt.ml [hunk 2]
```diff
@@ -88,30 +88,36 @@ let with_box_debug k = with_pp (Box_debug.with_box (fun fs -> eval fs k))
...
 let space_break =
+  (* a stack is useless here, this would require adding a unit parameter *)
+  let stack = "" in
   with_pp (fun fs ->
-      Box_debug.space_break fs ;
+      Box_debug.space_break ~stack fs ;
       Format_.pp_print_space fs () )
 
 let cut_break =
+  let stack = Box_debug.get_stack () in
```
Comment (Julow) [inline lib/Fmt.ml:106]: Should be the same as in the space case, no trace ? Same in `force_newline`.

File: lib/Fmt.ml [hunk 3]
```diff
@@ -88,30 +88,36 @@ let with_box_debug k = with_pp (Box_debug.with_box (fun fs -> eval fs k))
...
   with_pp (fun fs ->
-      Box_debug.break fs n o ;
+      Box_debug.break fs n o ~stack ;
       Format_.pp_print_break fs n o )
 
 let force_break = break 1000 0
 
 let space_break =
+  (* a stack is useless here, this would require adding a unit parameter *)
+  let stack = "" in
```
Comment (Julow) [inline lib/Fmt.ml:100]: Should be an option.

File: lib/Fmt.ml (hunk 2 above)
Comment (EmileTrotignon) [reply lib/Fmt.ml:106]: Yes this is true

File: lib/Conf.ml (hunk 1 above)
Comment (EmileTrotignon) [reply lib/Conf.ml:1405]: Enabling the stacktraces is a bit costly in terms of performances, and they have to be collected early. This allows to disable them when they are not going to be used, that is when there is no debug run that is going to be performed.

Comment (Julow) [reply lib/Conf.ml:1405]: There's already a bool variable for that: `Fmt.debug`. According to what you said, `enable_stacktraces` is unnecessary.

File: lib/Fmt.ml (hunk 2 above)
Comment (Julow) [reply lib/Fmt.ml]: There's still the issue in `force_newline`.

File: lib/Conf.ml (hunk 1 above)
Comment (EmileTrotignon) [reply lib/Conf.ml:1405]: No, fmt_debug is set much later

Comment (EmileTrotignon) [reply lib/Conf.ml:1405]: Box_debug.debug is set by the following function :
```
val with_box_debug : t -> t
```

When this function is called, the `t` you give as input is already generated, so its stack traces are either in memory or they are lost forever.

Comment (Julow) [reply lib/Conf.ml:1405]: Makes sense then!

Comment (Julow) [reply lib/Conf.ml:1405]: Though, to have clearer dependencies between the modules, I'd prefer if that variable was set from `Translation_unit`.

**Review by Julow — APPROVED** [review]
This is helping a lot, thanks!

**Pull request merged**